### DISCUSSION TITLE
Export and Import Template at Runtime

### DISCUSSION THREAD

**Discussion by maldworth** [QuestPDF/QuestPDF#350] [Ideas]
We have a use case where pdf layouts are customized per tenant, and they change over time. We don't want to re-deploy our web application, but we'd like to be able to update the invoice used. So is there any way you can build in a capability to "export" the document from the fluent api, so it could be saved (probably in some text markup/yaml template). Then our webapp at runtime, it would pull down the quest invoice template for the tenant, substitute in any dynamic text (quantities, address, sub totals, totals etc...), and then generate the PDF.

**Comment by maldworth** [comment]
@MarcinZiabek any thoughts? It's okay if you say "no plans", but at least it would let me know to move on and evaluate other libraries.

Thanks

**Reply by MarcinZiabek** [reply]
@maldworth Hi 😁 I am very sorry for the delayed response. Having so many notifications, I sometimes loose track of older questions. 

Being capable of replacing certain aspects of the application without redeploying everything - this sounds like a typical scenario for the microservice. Where one small application is responsible for PDF generation and it can be updated as a small part of entire system.

QuestPDF uses plain C# and therefore you can use various conditions to dynamically alter the document structure. This way, the document structure can be changed without the need of code deployment. Just different code paths are executedd.

In theory, you can try to serialize the QuestPDF.Document object (with all associated state and methods). I don't think it is a good idea though. http://mikehadlow.blogspot.com/2011/04/serializing-continuations.html

You can also implement a hod-reload approach. Where you deploy your PDF generation code as dll. Then the application can listen to any changes and reimport that dll after changes, without rerunning application. Some IoC containers may have this capability.

**Reply by maldworth** [reply]
Thanks for the response Marcin. Yeah, a microservice to perform the pdf generation can work, but it still requires technical people to make the change. We would like a marketing or communications team to make the change, and so changing code and checking into a repository is not possible. Thank you for all the work, Quest PDF is still useful for some of our use cases.

**Comment by DawidRobertNowak** [comment]
Hello,

We are closing this thread due to inactivity. Thank you for taking the time to reach out and contribute to the project — your feedback helps us improve QuestPDF.

Should you encounter any further problems, please don't hesitate to open a new issue.

Thank you!
